### ISSUE TITLE
Startup and liveness probes considered harmful

### ISSUE THREAD

**Issue by brancz** [prometheus-operator/prometheus-operator#7096]
### Component(s)

Prometheus, PrometheusAgent, AlertManager, ThanosRuler

### What is missing? Please describe.

We had some kubelets that were misbehaving and for some reason liveness probes were taking longer than usual and timing out, we're still investigating why exactly this was happening, but it begged the question for us: What is the purpose of these liveness and startup probes in the first place?

The typical startup and liveness probe argument is that it resolves an application's deadlocks, making it an auto-remediation of an issue. So in the absolute best-case scenario, these probes hide a real issue that should be investigated. However, in my soon 10 years of running Prometheus, not once was this appropriate, and we have countless examples of startup and liveness probes killing Prometheus servers when they shouldn't have (whether during WAL replay, or other random failures like this one).

We mitigated this by pausing the alertmanager and prometheus objects, removing startup and liveness probes and increased readiness probes' timeouts dramatically (to 30s). That resolved things.
Thinking about it on a higher level, I think the monitoring infrastructure simply shouldn't have liveness probes because, as this incident shows, there are many reasons why liveness probes fail that have nothing to do with the process. On top of that, everyone who is genuinely worried about the alerting pipeline not working should already have a real end-to-end test with the WatchDog alert.

Bottom line, I think startup and liveness probes are only ever doing harm and never good, I think we should remove them.

### Describe alternatives you've considered.

Using strategic merge patch and override things just for us. I think this is an improvement that everyone should get though.

## Environment
Kubernetes Version: GKE 1.31.2
Prometheus-Operator Version: v0.75.2

**Comment by brancz** [comment]
cc @simonpasquier

**Comment by nicolastakashi** [comment]
I tend to agree @brancz!
I already faced issues with probes filling process during long wal replays, or eventually healthz endpoint taking longer to respond, not meaning any real issue.

What I did back in time especially to the WAL replay is increase the startup probe but I agree this is a workaround :(

**Comment by brancz** [comment]
The way I see it, we should just remove them, and users who feel like there is value for these probes for them can add them via the strategic merge patch. I feel like this is the only option we have, because it's impossible to remove a probe using the strategic merge patch.

**Comment by simonpasquier** [comment]
Yeah, it's something that we've experienced too in the past. I would agree that liveness probes are not useful: I've never seen a case with a real deadlock which would be solved by restarting the container.
The startup probe is useful though because it ensures that you don't roll out a new version of one statefulset's pod before the other pod is ready (e.g. WAL replayed).
